### ISSUE TITLE
Enabling the Jupyter extension causes high CPU usage

### ISSUE THREAD

**Issue by ivanistheone** [microsoft/pylance-release#4022]
I'm on a Mac OS X 11.7.1, using vscode with latest Pylance.

When I enable the Jupyter extension (actually four of them get installed).
My CPU goes into 200%+ and doesn't stop due to the process `Code Helper (Plugin)`.

The Jupyter log doesn't say much:
```
Visual Studio Code (1.75.1, undefined, desktop)
Jupyter Extension Version: 2023.1.2010391206.
Python Extension Version: 2023.2.0.
Workspace folder /Users/ivan/Projects/Minireference/STATSbook/noBSstatsnotebooks
User belongs to experiment group 'jupyterTest'
User belongs to experiment group 'jupyterEnhancedDataViewer'
info 14:25:29.632: LSP Notebooks experiment is enabled
```

## Environment data

-   VS Code version: Version: 1.75.1 (Universal)
-   Jupyter Extension version: v2023.1.2010391206
-   Python Extension version: v2023.2.0
-   Pylance v2023.2.20
-   OS (Windows | Mac | Linux distro) and version: Mac OS X 11.7.1
-   Python and/or Anaconda version: Python 3.9.4
-   Type of virtual environment used: venv
-   Jupyter server running: not running at all...

## Expected behaviour

- Opening vscode to edit a python source file https://github.com/minireference/noBSstatsnotebooks/blob/main/notebooks/plot_helpers.py should not cause high CPU usage.

## Actual behaviour

- CPU goes to 200%+
- 

## Steps to reproduce:

1. Clone this repo https://github.com/minireference/noBSstatsnotebooks
   (contains lots of notebooks and 
2. Create a virtualenv called venv in the same folder
3. Activate the venv and pip install -r [requirements.txt](https://github.com/minireference/noBSstatsnotebooks/blob/main/requirements.txt)
4. In vscode with enabled Python (Pylance) and Juputer extensions
5. Open the file notebooks/plot_helpers.py for editing

## Logs

See this comment for detailed logs of Python and Python language server: https://github.com/microsoft/pylance-release/issues/3762#issuecomment-1426161232

- The Python log shows I have a bunch of virtualenvs all over the place and lots of errors when it tries to "inspect them"
- The Pylance log seems normal 

Let me know if I can do something to help track this down. For now I have disabled Jupyter extension since I use vscode mostly as code editor (source files only; no notebooks). BUt I can reliabrly reprdouce every time I turn on the Jupyter extension.

**Comment by rchiodo** [comment]
It looks to me like there are two different problems in this issue.

@ivanistheone is having high CPU usage with the Code plugin helper. This is likely the pylance language server (it runs with one of these helpers)

@stephinity is having high CPU usage with python executables running. Pylance doesn't run python executables except for a very short time on startup (to get sys.path), so I'm guessing that's something else. 

I'm going to concentrate on @ivanistheone's problem. 

For either problem the command line of the processes that are taking up so much CPU would be helpful.

For example on my machine, pylance is running under 'code.exe'. Its command line looks like so:

```
"C:\Users\aku91\AppData\Local\Programs\Microsoft VS Code\Code.exe" --ms-enable-electron-run-as-node c:\Users\aku91\.vscode\extensions\ms-python.vscode-pylance-2023.2.42\dist\server.bundle.js --cancellationReceive=file:2ad7fa4958b7080bf074ef4f4a05808b694983166b --node-ipc --clientProcessId=21456
```

**Comment by rchiodo** [comment]
Opening just the python file doesn't seem to cause a problem, but opening all of the notebooks in the same folder causes an out of memory error for pylance.

**Comment by rchiodo** [comment]
And without the jupyter extension enabled, pylance doesn't run out of memory.

**Comment by rchiodo** [comment]
Something that's weird, is pylance keeps getting file change events for the venv:

```
[Info  - 10:33:53 AM] (23856) LibraryFile: Received fs event 'change' for path 'd:\Source\Testing\noBSstatsnotebooks\venv\Lib\site-packages\numpy\testing\_private\decorators.py'
[Info  - 10:33:53 AM] (23856) SourceFile: Received fs event 'change' for path 'd:\Source\Testing\noBSstatsnotebooks\venv\Lib\site-packages\numpy\testing\_private\decorators.py'
[Info  - 10:33:53 AM] (23856) LibraryFile: Received fs event 'change' for path 'd:\Source\Testing\noBSstatsnotebooks\venv\Lib\site-packages\numpy\testing\_private\decorators.py'
[Info  - 10:33:53 AM] (23856) LibraryFile: Received fs event 'change' for path 'd:\Source\Testing\noBSstatsnotebooks\venv\Lib\site-packages\numpy\testing\_private\extbuild.py'
[Info  - 10:33:53 AM] (23856) SourceFile: Received fs event 'change' for path 'd:\Source\Testing\noBSstatsnotebooks\venv\Lib\site-packages\numpy\testing\_private\extbuild.py'
[Info  - 10:33:53 AM] (23856) LibraryFile: Received fs event 'change' for path 'd:\Source\Testing\noBSstatsnotebooks\venv\Lib\site-packages\numpy\testing\_private\extbuild.py'
``` 

This continually happen. Not sure why yet.

**Comment by debonte** [comment]
> pylance keeps getting file change events for the venv:

I think @bschnurr was seeing something like that recently?

**Comment by rchiodo** [comment]
I think the root cause of this problem is having multiple notebooks open. Pylance is not handling multiple notebooks in a very efficient manor. Each notebook basically gets its own copy of all of the 3rd party libraries loaded into memory. 

In the case above, I have 12 notebooks open which inside of Pylance creates 24 copies of every 3rd party library.

**Comment by stephinity** [comment]
An update from my side. The problem has "disappeared" with recent updates.

With updated extensions, running Jupyter notebooks with Python & Pylance extensions activated does NOT create high CPU usage anymore.

VS Code version: Code 1.76.0 (https://github.com/microsoft/vscode/commit/92da9481c0904c6adfe372c12da3b7748d74bdcb, 2023-03-01T10:22:44.506Z)

<html>
<body>


Extension|Author (truncated)|Version
-- | -- | --
python | ms- | 2023.4.0
vscode-pylance | ms- | 2023.3.10
jupyter | ms- | 2023.2.1000592019


</body>
</html>


@ivanistheone is your issue resolved also?

**Comment by ivanistheone** [comment]
Yes, I also found the issue is gone, but not in the (release) version:

```
GIVEN: 
 Pylance v2023.3.20  (release)
 Jupyter v2023.2.1200692131 (release)
THEN: high-CPU still present  EEEEEEEE
```

But since you mentioned might be fixed, I tried the pre-release version and issue is not present:

```
GIVEN:
 Jupyter v2023.3.1000711009 (pre-release)
 Pylance v2023.3.21  (pre-release)
THEN:
 No high-CPU issue   OK OK OK OK OK OK OK OK OK 
```

I also tried downgrading Pylance, while keeping Jupyter on pre-release and that also fixes the highCPU issue:


```
GIVEN:
 Pylance v2023.3.20  (release)
 Jupyter v2023.3.1000711009 (pre-release)
THEN:
 No high-CPU issue OK OK OK OK OK OK OK OK OK 
```

**Comment by ivanistheone** [comment]
@rchiodo  sorry for the delay in replying, here are the full command lines for my current setup when the high CPU issue is gone:

```
 Pylance v2023.3.20  (release)
 Jupyter v2023.3.1000711009 (pre-release)
Version: 1.76.1 (Universal)
Commit: 5e805b79fcb6ba4c2d23712967df89a089da575b
Date: 2023-03-08T16:32:09.831Z (1 day ago)
Electron: 19.1.11
Chromium: 102.0.5005.196
Node.js: 16.14.2
V8: 10.2.154.26-electron.0
OS: Darwin x64 20.6.0
Sandboxed: No
```

```
/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Renderer).app/Contents/MacOS/Code Helper (Renderer) --type=renderer --user-data-dir=/Users/ivan/Library/Application Support/Code --standard-schemes=vscode-webview,vscode-file --secure-schemes=vscode-webview,vscode-file --bypasscsp-schemes --cors-schemes=vscode-webview,vscode-file --fetch-schemes=vscode-webview,vscode-file --service-worker-schemes=vscode-webview --streaming-schemes --app-path=/Applications/Visual Studio Code.app/Contents/Resources/app --no-sandbox --no-zygote --node-integration-in-worker --lang=en-US --num-raster-threads=4 --enable-zero-copy --enable-gpu-memory-buffer-compositor-resources --enable-main-frame-before-activation --renderer-client-id=5 --launch-time-ticks=119794356470 --shared-files --field-trial-handle=1718379636,r,8736726377724882484,591659644482471017,131072 --disable-features=CalculateNativeWinOcclusion,SpareRendererForSitePerProcess --vscode-window-config=vscode:82aab5da-8e88-43d5-af1e-91a512bf3d3c --vscode-window-kind=shared-process

/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Renderer).app/Contents/MacOS/Code Helper (Renderer) --type=renderer --user-data-dir=/Users/ivan/Library/Application Support/Code --standard-schemes=vscode-webview,vscode-file --secure-schemes=vscode-webview,vscode-file --bypasscsp-schemes --cors-schemes=vscode-webview,vscode-file --fetch-schemes=vscode-webview,vscode-file --service-worker-schemes=vscode-webview --streaming-schemes --app-path=/Applications/Visual Studio Code.app/Contents/Resources/app --no-sandbox --no-zygote --enable-blink-features=HighlightAPI --lang=en-US --num-raster-threads=4 --enable-zero-copy --enable-gpu-memory-buffer-compositor-resources --enable-main-frame-before-activation --renderer-client-id=4 --launch-time-ticks=119792975053 --shared-files --field-trial-handle=1718379636,r,8736726377724882484,591659644482471017,131072 --disable-features=CalculateNativeWinOcclusion,SpareRendererForSitePerProcess --vscode-window-config=vscode:55f0713a-205e-4d89-81a5-aca41a082466

/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper.app/Contents/MacOS/Code Helper --type=utility --utility-sub-type=network.mojom.NetworkService --lang=en-US --service-sandbox-type=network --user-data-dir=/Users/ivan/Library/Application Support/Code --standard-schemes=vscode-webview,vscode-file --secure-schemes=vscode-webview,vscode-file --bypasscsp-schemes --cors-schemes=vscode-webview,vscode-file --fetch-schemes=vscode-webview,vscode-file --service-worker-schemes=vscode-webview --streaming-schemes --shared-files --field-trial-handle=1718379636,r,8736726377724882484,591659644482471017,131072 --disable-features=CalculateNativeWinOcclusion,SpareRendererForSitePerProcess --seatbelt-client=40
```





And for reference, here are the full commands when the high-CPU occurs:

Jupyter v2023.2.1200692131
Pylance v2023.3.20



```
/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (GPU).app/Contents/MacOS/Code Helper (GPU) --type=gpu-process --user-data-dir=/Users/ivan/Library/Application Support/Code --gpu-preferences=UAAAAAAAAAAgAAAIAAAAAAAAAAAAAAAAAABgAAAAAAAwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACgFAAAAAAAAKAUAAAAAAAC4AQAANgAAALABAAAAAAAAuAEAAAAAAADAAQAAAAAAAMgBAAAAAAAA0AEAAAAAAADYAQAAAAAAAOABAAAAAAAA6AEAAAAAAADwAQAAAAAAAPgBAAAAAAAAAAIAAAAAAAAIAgAAAAAAABACAAAAAAAAGAIAAAAAAAAgAgAAAAAAACgCAAAAAAAAMAIAAAAAAAA4AgAAAAAAAEACAAAAAAAASAIAAAAAAABQAgAAAAAAAFgCAAAAAAAAYAIAAAAAAABoAgAAAAAAAHACAAAAAAAAeAIAAAAAAACAAgAAAAAAAIgCAAAAAAAAkAIAAAAAAACYAgAAAAAAAKACAAAAAAAAqAIAAAAAAACwAgAAAAAAALgCAAAAAAAAwAIAAAAAAADIAgAAAAAAANACAAAAAAAA2AIAAAAAAADgAgAAAAAAAOgCAAAAAAAA8AIAAAAAAAD4AgAAAAAAAAADAAAAAAAACAMAAAAAAAAQAwAAAAAAABgDAAAAAAAAIAMAAAAAAAAoAwAAAAAAADADAAAAAAAAOAMAAAAAAABAAwAAAAAAAEgDAAAAAAAAUAMAAAAAAABYAwAAAAAAABAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAACAAAAEAAAAAAAAAAAAAAABwAAABAAAAAAAAAAAAAAAAgAAAAQAAAAAAAAAAAAAAAJAAAAEAAAAAAAAAAAAAAACwAAABAAAAAAAAAAAAAAAAwAAAAQAAAAAAAAAAAAAAAOAAAAEAAAAAAAAAAAAAAADwAAABAAAAAAAAAAAQAAAAAAAAAQAAAAAAAAAAEAAAACAAAAEAAAAAAAAAABAAAABwAAABAAAAAAAAAAAQAAAAgAAAAQAAAAAAAAAAEAAAAJAAAAEAAAAAAAAAABAAAACwAAABAAAAAAAAAAAQAAAAwAAAAQAAAAAAAAAAEAAAAOAAAAEAAAAAAAAAABAAAADwAAABAAAAAAAAAABAAAAAAAAAAQAAAAAAAAAAQAAAACAAAAEAAAAAAAAAAEAAAABwAAABAAAAAAAAAABAAAAAgAAAAQAAAAAAAAAAQAAAAJAAAAEAAAAAAAAAAEAAAACwAAABAAAAAAAAAABAAAAAwAAAAQAAAAAAAAAAQAAAAOAAAAEAAAAAAAAAAEAAAADwAAABAAAAAAAAAABwAAAAAAAAAQAAAAAAAAAAcAAAACAAAAEAAAAAAAAAAHAAAABwAAABAAAAAAAAAABwAAAAgAAAAQAAAAAAAAAAcAAAAJAAAAEAAAAAAAAAAHAAAACwAAABAAAAAAAAAABwAAAAwA
AAAQAAAAAAAAAAcAAAAOAAAAEAAAAAAAAAAHAAAADwAAABAAAAAAAAAACAAAAAAAAAAQAAAAAAAAAAgAAAACAAAAEAAAAAAAAAAIAAAABwAAABAAAAAAAAAACAAAAAgAAAAQAAAAAAAAAAgAAAAJAAAAEAA
AAAAAAAAIAAAACwAAABAAAAAAAAAACAAAAAwAAAAQAAAAAAAAAAgAAAAOAAAAEAAAAAAAAAAIAAAADwAAABAAAAAAAAAACgAAAAAAAAAQAAAAAAAAAAoAAAACAAAAEAAAAAAAAAAKAAAABwAAABAAAAAAAA
AACgAAAAgAAAAQAAAAAAAAAAoAAAAJAAAAEAAAAAAAAAAKAAAACwAAABAAAAAAAAAACgAAAAwAAAAQAAAAAAAAAAoAAAAOAAAAEAAAAAAAAAAKAAAADwAAAAgAAAAAAAAACAAAAAAAAAA= --shared-fil
es --field-trial-handle=1718379636,r,9901610019460239665,367251072297320160,131072 --disable-features=CalculateNativeWinOcclusion,SpareRendererForSitePerPr
ocess --seatbelt-client=39


/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.
framework/Helpers/chrome_crashpad_handler --no-rate-limit --monitor-self-annotation=ptype=crashpad-handler --database=/Users/ivan/Library/Application Support/Code/Crashpad --url=appcenter://code?aid=de75e3cc-e22f-4f42-a03f-1409c21d8af8&uid=60601833-f99c-4aab-8518-830f5172f346&iid=60601833-f99c-4aab-8518-830f5172f346&sid=60601833-f99c-4aab-8518-830f5172f346 --annotation=IsOfficialBuild=1 --annotation=_companyName=Microsoft --annotation=_productName=VSCode --annotation=_version=1.76.0 --annotation=exe=Electron --annotation=plat=OS X --annotation=prod=Electron --annotation=ver=19.1.11 --handshake-fd=27
iv

/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Plugin).app/Contents/MacOS/Code Helper (Plugin) --ms-enable-electron-run-as-node /Users/ivan/.vscode/extensions/ms-python.vscode-pylance-2023.3.20/dist/server.bundle.js --cancellationReceive=file:847009e96b604f241ffd7f3bdffbaed316591f36a4 --node-ipc --clientProcessId=11837
/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Renderer).app/Contents/MacOS/Code Helper (Renderer) --ms-enable-electron-run-as-node /Applications/Visual Studio Code.app/Contents/Resources/app/out/bootstrap-fork --type=fileWatcher
/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Renderer).app/Contents/MacOS/Code Helper (Renderer) --ms-enable-electron-run-as-node /Applications/Visual Studio Code.app/Contents/Resources/app/out/bootstrap-fork --type=ptyHost --logsPath /Users/ivan/Library/Application Support/Code/logs/20230310T072516
```

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.3.22, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023322-15-March-2023-PreRelease)
